src/app/clusters/zone-management-server/zone-util.h

```diff
@@ -0,0 +1,140 @@
+    {
+        return v1.x == v2.x && v1.y == v2.y;
+    }
+
+    // Method to check for self-intersecting zones
+    static bool IsZoneSelfIntersecting(const std::vector<TwoDCartesianVertexStruct> & vertices)
+    {
+        size_t vertexCount = vertices.size();
+
+        // Iterate through all pairs of non-adjacent segments
+        for (size_t i = 0; i < vertexCount; ++i)
+        {
+            TwoDCartesianVertexStruct p1 = vertices[i];
+            TwoDCartesianVertexStruct q1 = vertices[(i + 1) % vertexCount];
+
+            for (size_t j = i + 1; j < vertexCount; ++j)
+            {
+                TwoDCartesianVertexStruct p2 = vertices[j];
+                TwoDCartesianVertexStruct q2 = vertices[(j + 1) % vertexCount];
+
+                // Skip segments that share an endpoint (i.e., adjacent segments)
+                bool segments_are_adjacent = false;
+                if (AreTwoDCartVerticesEqual(p1, p2) || AreTwoDCartVerticesEqual(p1, q2) || AreTwoDCartVerticesEqual(q1, p2) ||
+                    AreTwoDCartVerticesEqual(q1, q2))
+                {
+                    segments_are_adjacent = true;
+                }
+
+                // Also a logical adjacency check for sequence
+                if ((j == (i + 1) % vertexCount) || (i == (j + 1) % vertexCount))
+                {
+                    segments_are_adjacent = true;
+                }
+
+                if (!segments_are_adjacent)
+                {
+                    if (DoSegmentsIntersect(p1, q1, p2, q2))
+                    {
+                        return true; // Found a self-intersection
+                    }
+                }
+            }
+        }
+
+        return false; // No self-intersection found
+    }
+
+private:
+    // Helper function: Check if point q lies on segment pr (assuming p, q, r are collinear)
+    // Checks if q is within the bounding box defined by p and r.
+    static bool OnSegment(const TwoDCartesianVertexStruct & p, const TwoDCartesianVertexStruct & q,
+                          const TwoDCartesianVertexStruct & r)
+    {
+        return q.x <= std::max(p.x, r.x) && q.x >= std::min(p.x, r.x) && q.y <= std::max(p.y, r.y) && q.y >= std::min(p.y, r.y);
+    }
+
+    // Helper function: Determine the orientation of the ordered triplet (p, q, r) of vertices.
+    // Employ the cross-product computation to find the direction of the Z axis.
+    // Using the right-hand rule Z > 0 means pointing upward from the 2D
+    // plane(counter-clockwise rotation) and Z < 0 means pointing downward into
+    // the 2D plane(clock-wise rotation).
+    // 0 --> p, q and r are collinear
+    // 1 --> Counterclockwise
+    // 2 --> Clockwise
+    static int Orientation(const TwoDCartesianVertexStruct & p, const TwoDCartesianVertexStruct & q,
+                           const TwoDCartesianVertexStruct & r)
+    {
+        long long val = static_cast<long long>((q.y - p.y) * (r.x - q.x)) - static_cast<long long>((q.x - p.x) * (r.y - p.y));
```

Comment:
This formula does not look right to me.  Where is it coming from?  This sort of thing should probably come with unit tests that would have ideally caught these problems.

It might help to explicitly document which vectors this is computing the cross-product of....